core/src/main/java/com/airbnb/aerosolve/core/transforms/FloatLabelTransform.java

```diff
@@ -0,0 +1,65 @@
+package com.airbnb.aerosolve.core.transforms;
+
+import com.airbnb.aerosolve.core.FeatureVector;
+import com.airbnb.aerosolve.core.util.Util;
+
+import java.util.Map;
+
+import com.typesafe.config.Config;
+
+/**
+ * Convert a Float feature to binary LABEL based on threshold value. This is intended for
+ * binary classification where LABEL is either -1 or 1. MergeStrategy describe how existing
+ * LABEL is treated.
+ */
+public class FloatLabelTransform implements Transform {
+  enum MergeStrategy {
+    IGNORE, // ignore existing label
+    KEEP_POSITIVE, // keep positive label
+    KEEP_NEGATIVE, // keep negative label
+    KEEP_ALL // preserve existing label (only replace if observation is un-labeled
+  }
+
+  private String fieldName1;
+  private String key1;
+  private double threshold;
+  private MergeStrategy mergeStrategy;
+
+  @Override
+  public void configure(Config config, String key) {
+    fieldName1 = config.getString(key + ".field1");
+    key1 = config.getString(key + ".key1");
+    threshold = config.getDouble(key + ".threshold");
+    mergeStrategy = MergeStrategy.valueOf(config.getString(key + ".merge").toUpperCase());
+  }
+
+  @Override
+  public void doTransform(FeatureVector featureVector) {
+    Map<String, Double> labelFeature = Util.getOrCreateFloatFeature("LABEL", featureVector.floatFeatures);
+
+    Double label = labelFeature.get("");
+
+    if (mergeStrategy == MergeStrategy.KEEP_ALL && label != null) return;
+
+    Map<String, Double> floatFeature = featureVector.floatFeatures.get(fieldName1);
+    if (floatFeature != null) {
+      Double featureValue = floatFeature.get(key1);
+      if (featureValue != null) {
+        double newLabel = featureValue >= threshold ? 1 : -1;
+
+        if (
```

Comment:
I don't insist on changing it to separate variables. But I would like to share my experience on this kind of 'optimization'. Many people are taught about this kind of short-circuit could produce big performance gain. But in practice, it never worked.

There are many reasons behind. But in general, it's because any program has much larger overhead to run which usually involve many I/O. That overhead will marginalize the performance gain from small memory operation.

For this particular case, I actually ran an experiment to demonstrate this effect. I think the two programs should have more pronounced performance difference from the short-circuiting than our real spark jobs. 

The short-circuiting version:

```scala
object ShortCut {
  def main(args: Array[String]) {
    var count = 0
    val r = scala.util.Random
    for (i <- 1 to 1000) {
      if (r.nextDouble >= 0.5 || r.nextDouble >= 0.5 || r.nextDouble >= 0.5 || r.nextDouble >= 0.5) {
        count = count + 1
      }
    }
  }
}
```

The non-short-circuiting version:

```scala
object NonShortCut {
  def main(args: Array[String]) {
    var count = 0
    val r = scala.util.Random
    for (i <- 1 to 1000) {
      val a = r.nextDouble
      val b = r.nextDouble
      val c = r.nextDouble
      val d = r.nextDouble
      //println("%.3f, %.3f, %.3f, %.3f".format(a, b, c, d))
      if (a >= 0.5 || b >= 0.5 || c >= 0.5 || d >= 0.5) {
        count = count + 1
      }
    }
  }
}
```

Compiling and run each ten times, the elapsed time are almost always the same. Below is one instance of the experiment.

```
 $ time for i in `seq 1 10`; do scala NonShortCut; done
941
951
932
943
938
928
943
928
928
930

real	0m3.284s
user	0m3.117s
sys	0m0.976s

 $ time for i in `seq 1 10`; do scala ShortCut; done
947
945
928
939
938
940
950
942
941
942

real	0m3.300s
user	0m3.137s
sys	0m0.973s

```

Similar proof can be found very widely. For example, Python is considered slow(indeed in lots of context), but training with Sklearn is very fast because all key computation like matrix multiply is done in Fortune or C. The performance issue in user program, like the Python for-loop or branching, have no effect in any real program.

My two cents.